src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs

```diff
@@ -0,0 +1,435 @@
+                // start on the current line. We do not need to check that this posteb
+                // comes after the preeb because the preeb's prior content has already
+                // been validated to be whitespace.
+                do
+                {
+                    ReadOnlySpan<char> postebLine = pemData[postebLineRange];
+                    int postebIndex = postebLine.IndexOf(posteb);
+
+                    if (postebIndex == -1)
+                    {
+                        continue;
+                    }
+
+                    (int postebOffset, _) = postebLineRange.GetOffsetAndLength(pemData.Length);
+                    int postebEndIndex = postebOffset + postebIndex + posteb.Length;
+                    Range location = preebStartIndex..postebEndIndex;
+                    Range content = (endLabelIndex + Ending.Length)..(postebOffset + postebIndex);
+
+                    if (!postebLine[(postebIndex + posteb.Length)..].IsWhiteSpace())
+                    {
+                        break;
+                    }
+
+                    if (IsValidBase64(pemData, content, out Range base64range, out int decodedBase64Size))
+                    {
+                        fields = new PemFields(labelRange, base64range, location, decodedBase64Size);
+                        return true;
+                    }
+                    break;
+                }
+                while (TryReadNextLine(pemData, ref postebLinePosition, out postebLineRange));
+            }
+
+            fields = default;
+            return false;
+        }
+
+        private static bool IsValidLabel(ReadOnlySpan<char> data)
+        {
+            static bool IsLabelChar(char c) => c >= 0x21 && c <= 0x7e && c != '-';
+
+            if (data.Length == 0)
+                return true;
+
+            // First character of label must be a labelchar
+            if (!IsLabelChar(data[0]))
+                return false;
+
+            for (int index = 1; index < data.Length; index++)
+            {
+                char c = data[index];
+                if (!IsLabelChar(c) && c != ' ' && c != '-')
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private static bool IsValidBase64(
+            ReadOnlySpan<char> data,
+            Range content,
+            out Range base64range,
+            out int decodedSize)
+        {
+            static bool IsBase64Char(char c) =>
+                (c >= '0' && c <= '9') ||
+                (c >= 'A' && c <= 'Z') ||
+                (c >= 'a' && c <= 'z') ||
+                c == '+' || c == '/' || c == '=';
+
+            int base64chars = 0;
+            int precedingWhiteSpace = 0;
+            int trailingWhiteSpace = 0;
+            (int offset, int length) = content.GetOffsetAndLength(data.Length);
+            for (int index = offset; index < offset + length; index++)
```

Comment:
I don't know how important perf is here, but I suspect this validation is quite slow (and doing the actual encoding using the Base64 APIs built on top of avx2/Ssse3 might actually be faster).

Alternatively, we could use some of the Base64 implementation as inspiration to make this method faster.

How large can `length` be?

Reply:
> but I suspect this validation is quite slow

Yeah, it occurred to me but when I looked around for inspiration such as at https://github.com/dotnet/runtime/blob/aac8f91cc5510b3d271d86eef616b6007b6e6631/src/libraries/System.Private.CoreLib/src/System/Convert.cs#L2820

I see it still doing character-by-character examination to account for various whitespace (albeit with unsafe tricks), just that the method has a looser definition of whitespace than the RFC.

Is there a better place to be looking here for this?

> How large can length be?

It could be very large, but the typical use-case for PEM is PKI and CMS. CMS can be several thousand characters.

Reply:
Validation could be moved to `System.Buffers.Text.Base64`.  It is already vectorized, and contains a (private) decoding map that can be used to perform an efficient validation with vector instructions, with a scalar fallback.   The functions that decode using AVX2 and SSSE3 already check for invalid input before decoding, so it should be trivial.

Reply:
We would need to make that validation "whitespace" aware. The existing `S.B.T.Base64` APIs do not support ignoring whitespace characters.

Reply:
FWIW, something like this (which includes the work of actually decoding) is ~2x faster than doing the linear validation (for a ~1k buffer without any leading/trailing whitespace). I am not necessarily suggesting that this is the way to go (especially given the unnecessary work it does and is no where near "optimal"), but at the very least it avoids having to provide your own custom logic for the base64 validation. Ideally, this type of API is built-in to either the `Convert` or `Base64` classes.

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.19041
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-alpha1-015914
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.19.56303, CoreFX 5.0.19.56306), X64 RyuJIT
  Job-XEOKAL : .NET Core 5.0.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  Toolchain=CoreRun  MaxIterationCount=10  
MinIterationCount=5  WarmupCount=3  

```
|           Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Ratio | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ValidateOriginal | 6.308 us | 0.1829 us | 0.1088 us | 6.316 us | 6.186 us | 6.548 us |  1.00 |     - |     - |     - |         - |
|      ValidateNew | 3.290 us | 0.0724 us | 0.0431 us | 3.302 us | 3.214 us | 3.352 us |  0.52 |     - |     - |     - |         - |


```C#
private static readonly char[] s_whitespace = new char[] { ' ', '\t', '\r', '\n' };
private static readonly char[] s_base64 = new char[]
{
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h' ,'i', 'j',
    'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r' ,'s', 't',
    'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B' ,'C', 'D',
    'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L' ,'M', 'N',
    'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V' ,'W', 'X',
    'Y', 'Z', '+', '/'
};

private bool IsValidBase64(out int decodedSize)
{
    ReadOnlySpan<char> input = _base64String.AsSpan();

    int startIndex = input.IndexOfAny(s_base64);

    if (startIndex == -1)
    {
        startIndex = 0;
    }

    input = input.Slice(startIndex);

    int finalIndex = input.LastIndexOfAny(s_whitespace);

    if (finalIndex == -1)
    {
        finalIndex = input.Length;
    }

    input = input.Slice(0, finalIndex);

    byte[] buffer = ArrayPool<byte>.Shared.Rent(Base64.GetMaxDecodedFromUtf8Length(input.Length));

    bool result = Convert.TryFromBase64Chars(input, buffer, out decodedSize);
    if (!result)
    {
        decodedSize = 0;
    }

    ArrayPool<byte>.Shared.Return(buffer, clearArray: true);

    return result;
}
```